File: java-service/src/main/java/com/flashsale/api/service/SpuService.java
package com.flashsale.api.service;

import com.flashsale.api.dto.SpuDtos.SpuCreateDto;
import com.flashsale.api.dto.SpuDtos.SpuResponseDto;
import com.flashsale.api.dto.SpuDtos.SpuUpdateDto;
import com.flashsale.api.entity.Spu;
import com.flashsale.api.mapper.SpuMapper;
import com.flashsale.api.repository.SpuRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;
import java.util.UUID;

@Service
@Transactional
public class SpuService {

    private static final Logger logger = LoggerFactory.getLogger(SpuService.class);

    private final SpuRepository spuRepository;
    private final SpuMapper spuMapper;

    @Autowired
    public SpuService(SpuRepository spuRepository, SpuMapper spuMapper) {
        this.spuRepository = spuRepository;
        this.spuMapper = spuMapper;
    }

    @Transactional(readOnly = true)
    public Page<SpuResponseDto> getAllSpus(int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
        Page<Spu> spus = spuRepository.findAll(pageable);
        return spus.map(spuMapper::toResponseDto);
    }

    @Transactional(readOnly = true)
    public Optional<SpuResponseDto> getSpuById(UUID id) {
        return spuRepository.findById(id)
                .map(spuMapper::toResponseDto);
    }

    @Transactional(readOnly = true)
    public Optional<SpuResponseDto> getSpuBySlug(String slug) {
        return spuRepository.findBySlug(slug)
                .map(spuMapper::toResponseDto);
    }

    public SpuResponseDto createSpu(SpuCreateDto createDto) {
        // Check if slug already exists
        if (spuRepository.existsBySlug(createDto.getSlug())) {
            throw new IllegalArgumentException("SPU with slug '" + createDto.getSlug() + "' already exists");
        }

        Spu spu = spuMapper.toEntity(createDto);
        Spu savedSpu = spuRepository.save(spu);

        logger.info("Created SPU with ID: {} and name: {}", savedSpu.getId(), savedSpu.getName());

        return spuMapper.toResponseDto(savedSpu);
    }

    public Optional<SpuResponseDto> updateSpu(UUID id, SpuUpdateDto updateDto) {
        return spuRepository.findById(id)
                .map(existingSpu -> {
                    // Check slug uniqueness if being updated
                    if (updateDto.getSlug() != null && 
                        !updateDto.getSlug().equals(existingSpu.getSlug()) &&
                        spuRepository.existsBySlug(updateDto.getSlug())) {
                        throw new IllegalArgumentException("SPU with slug '" + updateDto.getSlug() + "' already exists");
                    }

                    spuMapper.updateEntityFromDto(updateDto, existingSpu);
                    Spu updatedSpu = spuRepository.save(existingSpu);

                    logger.info("Updated SPU with ID: {}", updatedSpu.getId());

                    return spuMapper.toResponseDto(updatedSpu);
                });
    }

    public boolean deleteSpu(UUID id) {
        if (spuRepository.existsById(id)) {
            spuRepository.deleteById(id);
            logger.info("Deleted SPU with ID: {}", id);
            return true;
        }
        return false;
    }
}

File: variant-x/csharp-service/Services/SpuService.cs
using AutoMapper;
using FlashSale.Api.Data;
using FlashSale.Api.DTOs;
using FlashSale.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace FlashSale.Api.Services;

public class SpuService : ISpuService
{
    private readonly FlashSaleDbContext _context;
    private readonly IMapper _mapper;
    private readonly ILogger<SpuService> _logger;

    public SpuService(FlashSaleDbContext context, IMapper mapper, ILogger<SpuService> logger)
    {
        _context = context;
        _mapper = mapper;
        _logger = logger;
    }

    public async Task<IEnumerable<SpuResponseDto>> GetAllAsync(int skip = 0, int take = 100)
    {
        var spus = await _context.Spus
            .OrderByDescending(s => s.CreatedAt)
            .Skip(skip)
            .Take(take)
            .ToListAsync();

        return _mapper.Map<IEnumerable<SpuResponseDto>>(spus);
    }

    public async Task<SpuResponseDto?> GetByIdAsync(Guid id)
    {
        var spu = await _context.Spus.FindAsync(id);
        return spu == null ? null : _mapper.Map<SpuResponseDto>(spu);
    }

    public async Task<SpuResponseDto> CreateAsync(SpuCreateDto dto)
    {
        // Check if slug already exists
        if (await _context.Spus.AnyAsync(s => s.Slug == dto.Slug))
        {
            throw new InvalidOperationException("SPU with this slug already exists");
        }

        var spu = _mapper.Map<Spu>(dto);
        _context.Spus.Add(spu);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Created SPU {SpuId} with name {Name}", spu.Id, spu.Name);
        
        return _mapper.Map<SpuResponseDto>(spu);
    }

    public async Task<SpuResponseDto?> UpdateAsync(Guid id, SpuUpdateDto dto)
    {
        var spu = await _context.Spus.FindAsync(id);
        if (spu == null) return null;

        // Check slug uniqueness if being updated
        if (!string.IsNullOrEmpty(dto.Slug) && dto.Slug != spu.Slug)
        {
            if (await _context.Spus.AnyAsync(s => s.Slug == dto.Slug))
            {
                throw new InvalidOperationException("SPU with this slug already exists");
            }
        }

        _mapper.Map(dto, spu);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Updated SPU {SpuId}", spu.Id);
        
        return _mapper.Map<SpuResponseDto>(spu);
    }

    public async Task<bool> DeleteAsync(Guid id)
    {
        var spu = await _context.Spus.FindAsync(id);
        if (spu == null) return false;

        _context.Spus.Remove(spu);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Deleted SPU {SpuId}", id);
        
        return true;
    }
}
